### PR TITLE
Add support for Intel Centrino Wireless-N wireless network devices

### PR DESCRIPTION
**Pull request by pigworlds**
Port over iwn driver from OpenBSD project.

The iwn driver provides support for Intel Wireless WiFi Link 4965/5000/1000 and Centrino Wireless-N 1000/2000/6000 Series PCIe Mini Card network adapters. *Untested

### CODE REVIEW

**Comment by pigworlds** [comment]
@zxystd Thank you create this great project.
As inspired, this work adds iwn driver ported from OpenBSD project. This adds to support more down-level devices.

**Comment by pigworlds** [comment]
I've tested on Intel Centrino Advanced-N 6205 ONLY. For anyone is interested testing this driver before merge, you can try it here.

[AirportItlwm-Catalina.zip](https://github.com/OpenIntelWireless/itlwm/files/5551959/AirportItlwm-Catalina.zip)
[AirportItlwm-Big Sur.zip](https://github.com/OpenIntelWireless/itlwm/files/5551961/AirportItlwm-Big.Sur.zip)
[itlwm.zip](https://github.com/OpenIntelWireless/itlwm/files/5551964/itlwm.zip) with HeliPort

**Comment by zxystd** [comment]
@pigworlds Thanks for your work. Have you ever do benchmark Speedtest on this iwn driver? TBH, we don't have plan to support it, because we don't have these devices and don't have too much effort to maintain it...

**Comment by pigworlds** [comment]
The speed of this iwn driver is about 55mbps down, and 20mbps up when measured using iperf3 in my environment. This is consistent with Speedtest on the download speed, and upload speed is limited by my internet provider. This isn't bad either compare to the speed listed in FAQ.

I fortunately and have a laptop with this Intel Centrino Advanced-N 6205. Unfortunately the wireless card on the laptop is unreplaceable. I have no other option to use wireless on the laptop other than use it with iwn driver. Most likely I'm able to support this iwn driver until the the laptop stop working.

**Comment by zxystd** [comment]
Ok. I see.

File: itlwm.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -886,10 +912,34 @@
 				F8A68DDB2482242E007EB376 /* iwlwifi-cc-a0-48.ucode */,
 				F8A68DD72482224F007EB376 /* iwlwifi-Qu-b0-hr-b0-48.ucode */,
 				F8A68DD92482241E007EB376 /* iwlwifi-QuZ-a0-hr-b0-48.ucode */,
+				17FD7F43255E4EB300611406 /* iwn-100 */,
+				17FD7F3E255E4EB300611406 /* iwn-105 */,
```
Comment (zxystd) [inline itlwm.xcodeproj/project.pbxproj:916]: iwn-100 is not in use, maybe should delete it.

File: itlwm/hal_iwn/ItlIwn.cpp [hunk 2]
```diff
@@ -0,0 +1,7306 @@
...
+is5GBandSupport()
+{
+    return com.sc_flags & IWN_FLAG_HAS_5GHZ;
+}
+
+int ItlIwn::
+getTxNSS()
+{
+    // return !com.sc_nvm.sku_cap_mimo_disable ? (iwn_mimo_enabled(&com) ? 2 : 1) : 1;
+    return 1;
```
Comment (zxystd) [inline itlwm/hal_iwn/ItlIwn.cpp:207]: should not hard coded to 1, some devices may support MIMO, should return the value of `sc->ntxchains`

File: itlwm/hal_iwn/ItlIwn.cpp [hunk 3]
```diff
@@ -0,0 +1,7306 @@
...
+
+    sc->sc_txtap_len = sizeof sc->sc_txtapu;
+    sc->sc_txtap.wt_ihdr.it_len = htole16(sc->sc_txtap_len);
+    sc->sc_txtap.wt_ihdr.it_present = htole32(IWN_TX_RADIOTAP_PRESENT);
+}
+#endif
+
+//int ItlIwn::
+//iwn_detach(struct device *self, int flags)
+//{
```
Comment (zxystd) [inline itlwm/hal_iwn/ItlIwn.cpp:742]: what happened?

File: itlwm/hal_iwn/ItlIwn.cpp [hunk 4]
```diff
@@ -0,0 +1,7306 @@
...
+int ItlIwn::
+iwn_intr(OSObject *object, IOInterruptEventSource* sender, int count)
+{
+    ItlIwn *that = (ItlIwn*)object;
+    struct iwn_softc *sc = &that->com;
+    struct _ifnet *ifp = &sc->sc_ic.ic_if;
+    uint32_t r1, r2, tmp;
+
+    /* Disable interrupts. */
+    IWN_WRITE(sc, IWN_INT_MASK, 0);
```
Comment (zxystd) [inline itlwm/hal_iwn/ItlIwn.cpp:3271]: don't need to disable interrupt again.

File: itlwm.xcodeproj/project.pbxproj (hunk 1 above)
Comment (pigworlds) [reply itlwm.xcodeproj/project.pbxproj:914]: Removed. This iwn-100 file come from the firmware package, and apparently it is not being used.

File: itlwm/hal_iwn/ItlIwn.cpp (hunk 2 above)
Comment (pigworlds) [reply itlwm/hal_iwn/ItlIwn.cpp]: Thanks a lot to point out. I understand incorrectly the MIMO was introduced only to 802.11ac, and it was actually in the N standard. I can find the logs indicate 2T2R, so the TxNSS should be 2.
I also noticed the upload speed increased to 30mbps while test using iperf3 after this change.

File: itlwm/hal_iwn/ItlIwn.cpp (hunk 3 above)
Comment (pigworlds) [reply itlwm/hal_iwn/ItlIwn.cpp]: Removed the unused code.

File: itlwm/hal_iwn/ItlIwn.cpp (hunk 4 above)
Comment (pigworlds) [reply itlwm/hal_iwn/ItlIwn.cpp]: Yeah, just found it was the interrupt was already disabled.

**Comment by zxystd** [comment]
Ok. I don't have device to test it, no more suggestions now, if you find some issues when you are using it, please let me know. Thank you for the contribution again.

**Comment by codic12** [comment]
I'm using the alpha that has this PR merged and I see all my networks. However when I select one and key it in, it connects me, but shows an exclamation mark on the wifi sign and does not let me use Safari. This is all on the recovery USB, btw.

**Comment by pigworlds** [comment]
@codic12 Thanks try this out and share the information. Which wifi card are you using? What method of loading the kext? Through OC injection, Clover injection or kextload? Will you able to grab the logs? `dmesg | grep -i itlwm` should be ok to get last few lines of logs if it is difficulty get full logs in the recovery mode.

During my testing I do see similar issue mostly in the recovery mode. A cycle of turn off wifi and turn on fix this most of time.

**Comment by codic12** [comment]
@pigworlds Hi, same wifi card as you: Intel Centrino Advanced-N 6205. 
I'm loading the kext through OpenCore (copying to EFI/OC/Kexts) and adding to Config.plist.
I'll try grabbing logs with dmesg (do that often on my Linux setup, did not know macos had it!).

And I'll try turning it off and on too.

**Comment by codic12** [comment]
Ok, so there was some messages about not recieveing something. However I turned it off and on, and wow it works! I am on the live usb right now, thank you so much!

**Comment by pigworlds** [comment]
@codic12 Glad it works for you. :-) Once you have the full system setup, please report back how it works.

Please note, there are 2 known minor when you inject through OC bootloader:
1. After boot the system wireless already configured, you may hit a wifi glitch for a few seconds around 1 minute after logon. The logs indicates it hits the iwn_watchdog timeout. It will automatically restart the wifi card and reconnect. You don't need to take any action but just the glitch.
2. Sometimes you may unable to connect to the wireless in Recovery mode (the issue just reported). Normally a turn off/turn-on cycle in the wifi menu works.

Also note, the iwn driver seems doesn't support MIMO (only support up to SISO MCS 7) and no support 40Mhz channel yet. Please run a speedtest to check the performance of it.

**Comment by codic12** [comment]
Hi @pigworlds,
AirportItlwm was enough to get my Hackintosh installed in under 30 minutes, and it's running very smoothly. However I cannot access wifi, there is no Airport menu in the top bar at all!

(It's been much more than a minute.)

Is there any way to fix this?

**Comment by codic12** [comment]
Oh, and a wifi card does not show in Network Preferences either. Only Ethernet&Modem.

**Comment by ErrorErrorError** [comment]
> Oh, and a wifi card does not show in Network Preferences either. Only Ethernet&Modem.

See if you can add it by pressing the "+" icon on the bottom left in `Network Preferences` and find interface `Wireless` or `Wi-Fi` .

**Comment by codic12** [comment]
No, no interface like that.
Dunno if itlwm is even started, since grep'ing for it in dmesg does not show it `dmesg | grep -i itlwm`

**Comment by pigworlds** [comment]
This sounds the kext is not injected properly. Can you run this to check if it is loaded?
kextstat |grep -i itlwm

Can you manually run this to load to see if it works?
sudo kextload path/to/AirportItlwm.kext

**Comment by codic12** [comment]
I will try those commands. However, I cannot figure out how to force the kernel (cannot find that part of the manual) to load IO80211Family...?

**Comment by codic12** [comment]
It was just my own stupidity indeed, normal itlwm works perfectl! Speeds are very acceptable, slightly lower than Linux.

**Pull request merged**